### DISCUSSION TITLE
Best practice: dynamic Charka props based on boolean component prop

### DISCUSSION THREAD

**Discussion by TylerAPfledderer** [chakra-ui/chakra-ui#4737] [Q&A] [answered]
Hey all!

Say there is a component boolean prop that if it returns true, will change or add some styles in the component. 
And sometimes, if the prop is false, then the Chakra prop doesn't need value. One would assume a short-circuit operation would suffice.

In the component:
```
const Example = ( isProp=false ) => {
  return (
    <Box background={isProp && "black"}>
    {/* Rest of the component */}
```

And calling the component:
```
{/* Parent component jsx above...*/}
  <Example />
  <Example isProp />
{/* Parent component jsx below...*/}
```

But you can't short circuit because Chakra props can not accept a boolean that returns if `isProp` is `false`. 

A ternary operator can only work here, such as `background={isProp ? "black" : ""}`

This can get a little hairy when applied partially to a breakpoint object, such as 
`marginX={{base: isProp ? "10px" : "", md: "20px"}}`

But if there are multiple Chakra props that I have to do this pattern with, it seems like that can get pretty messy.

What is the best practice in handling a component prop that impacts multiple Chakra Props, especially Chakra props that should only contain either one value or no value?

**Edit:** the example and issue are in relation to using Typescript. See my [initial response to primos63](https://github.com/chakra-ui/chakra-ui/discussions/4737#discussioncomment-1387192).

Thank you!

**Comment by primos63** [comment]
Hopefully, I've understood your question, if there's something I've missed please let me know.

The Example component is incorrect. Components are passed `props` which can be destructured.
```jsx
const Example = ( {isProp=false, ...props} ) => {
  return (
    <Box background={isProp && "gray.400"} {...props}>
      Gray.400 if true else default background color
    </Box>
  )
}
```
and
```jsx
<Example />
<Example isProp />
```
work as expected. You can use short circuiting as properties with undefined are removed.

For the breakpoint, I don't think you would have a case where `base` would be undefined as that wouldn't make sense, but you could use it for other breakpoints. Even here short circuting works.

Here's a link to my [CodeSandbox](https://codesandbox.io/s/relaxed-goodall-1up04?file=/src/App.js) which demonstrates using a boolean prop and short circuiting.

**Reply by TylerAPfledderer** [reply]
Hey @primos63!

Thank you for your response and apologies for the delay. 

Thank you for the correction in the syntax for the props in my example. What I also failed to mention is I am also using Typescript. 🤦‍♂️

Passing in a short circuit operation to a Chakra prop, I receive a type error like this:
```
background={isProp && 'gray.400'}

// Typescript error
Type 'false | "gray.400"' is not assignable to type 'ResponsiveValue<Union<"gray.400" | Color | "current" | "whiteAlpha.50" | "whiteAlpha.100" | "whiteAlpha.200" | "whiteAlpha.300" | "whiteAlpha.400" | "whiteAlpha.500" | "whiteAlpha.600" | ... 171 more ... | "telegram.900">> | undefined'.
  Type 'false' is not assignable to type 'ResponsiveValue<Union<"gray.400" | Color | "current" | "whiteAlpha.50" | "whiteAlpha.100" | "whiteAlpha.200" | "whiteAlpha.300" | "whiteAlpha.400" | "whiteAlpha.500" | "whiteAlpha.600" | ... 171 more ... | "telegram.900">> | undefined'.ts(2322)
```

**Comment by primos63** [comment, marked as answer]
Hey Tyler,

No worries. Ah TypeScript! If it says no, then it must be for your own good. :laughing: 

I think you want your prop to be more like an attribute where you want to test if it exists rather than if it's set to true or false. If I change `isProp` to `isGrayed` in `Example.tsx` where grayed is an optional property which can be either `true` or `undefined` (the default)
```jsx
import { Box } from "@chakra-ui/react";

type ExampleProps = {
  isGrayed?: true | undefined;
};

export const Example = ({ isGrayed = undefined, ...props }: ExampleProps) => {
  return (
    <Box background={isGrayed && "gray.400"} {...props}>
      Gray.400 if isGrayed exists
    </Box>
  );
};
```
then I would be able to set the background color with
```jsx
import { Container } from "@chakra-ui/react";
import { Example } from "./Example";

export default function App() {
  return (
    <Container>
      <Example />
      <Example isGrayed />
    </Container>
  );
}
```

This is something like what Chakra does for data attributes like `data-checked` which comes from `isChecked`.
Example using this method
```jsx
import { Box } from "@chakra-ui/react";

type Booleanish = true | undefined

const dataAttr = (condition: boolean | undefined) => 
  (condition ? true : undefined) as Booleanish

type ExampleProps = {
  isGrayed?: boolean;
};

export const Example = ({ isGrayed = undefined, ...props }: ExampleProps) => {
  return (
    <Box background={dataAttr(isGrayed) && "gray.400"} {...props}>
      Gray.400 if isGrayed exists
    </Box>
  );
};
```
and App
```jsx
import { Container } from "@chakra-ui/react";
import { Example } from "./Example";

export default function App() {
  return (
    <Container>
      <Example />
      <Example isGrayed />
      <Example isGrayed={false} />
    </Container>
  );
}
```
giving the result

![image](https://user-images.githubusercontent.com/2768849/134842775-ca93cc94-4ac0-4f32-9d41-98a74f406a5e.png)

**Reply by TylerAPfledderer** [reply]
Ok! What I am learning here is that it's ok to override the type error by declaring the union `true | undefined` of a certain boolean prop if the only error I will get is the possible inclusion of `false` which has no real negative effect when compiling the Chakra prop.

Am I understanding that correctly?
